### ISSUE TITLE
Suggestion: Delete orphaned dependencies when uninstalling

### ISSUE THREAD

**Issue by Alhadis** [Homebrew/brew#784]
I'd submit a PR for this, but my Ruby knowledge is crap.

It'd be nice if `brew uninstall` had a `--prune` option to kill off any dependencies that were made leaves by the removal. Some formulae have tonnes of dependencies, which might only ever be installed just to support that one program. Examples that come to mind are `ffmpeg`, `inkscape`, and practically any hefty program associated with GNOME toolkit.

Of course, after a while, it becomes impossible to remember *why* something was installed in the first place (particularly if you have hundreds of brews...). The idea for a `brew rm -rf` feature came from a [note](https://github.com/Alhadis/.files/commit/a6f8c1e7dcb3c7c6a3fd5b9589b484daedb08537#diff-7801e6443065d3e0c41018935a0bfff3L5) I left myself reminding me to uninstall unneeded formulae if I removed Inkscape.

**Comment by jawshooah** [comment]
Check out [`brew rmtree`](https://github.com/beeftornado/homebrew-rmtree).

**Comment by Alhadis** [comment]
I considered writing a shell-script to do this myself, but thought the better of it upon realising what's going to help more people in the long-term. :p

**Comment by MikeMcQuaid** [comment]
> I'd submit a PR for this, but my Ruby knowledge is crap.

Just FYI:
- stuff like this should be in https://github.com/Homebrew/brew-evolution
- the first Ruby I ever wrote was for Homebrew. I'm now writing Ruby full-time at GitHub

The best current solution to this is probably using https://github.com/Homebrew/homebrew-bundle to install/remove just the stuff you want. We'll need an implementation of https://github.com/Homebrew/brew-evolution/blob/master/003-install-bottles-of-arbitrary-versions.md#detailed-design first where installations record their dependencies.

**Comment by Alhadis** [comment]
Ah right, sorry. My bad. :(

Thanks!

**Comment by MikeMcQuaid** [comment]
No need to apologise! Mostly just closing this out because it's blocked on a few other things.

**Comment by hyiltiz** [comment]
This should really be supported by default (as opposed to via an external command), and arguably should be the default behavior for `brew uninstall`. It is the job of any decent package manager to keep track of dependencies installed along a package, and providing an option to remote all those useless orphans once they are no longer needed. Otherwise, we'll be stuck in a system that collects trash over time, and requires a fresh re-install (as if it was Windows).
